### PR TITLE
[DO-NOT-MERGE] sql: add a new index hint NO_SCAN to reject full table scans

### LINKED ISSUE #13969: sql: allow rejecting a query if a full table scan is involved
It would be great to support SQL syntax that causes a query to be rejected/fail if it would cause a full table scan to be performed.

For production use cases, in many situations it's always wrong to perform a table scan. This is often known at the time the code is written that executes queries. Yet things like forgetting to add an index or indexing the wrong thing or using slightly incorrect WHERE clauses can happen. It's generally useful for a database to be able to operate in a mode where such queries are rejected, rather than potentially taking down the database and impacting other traffic.

### PR DESCRIPTION
**Pull request by knz**
Prior to this patch a client could already indicate via "index hints"
how tables should be used, for example `SELECT * FROM
kv@{FORCE_INDEX=myindex}` or `SELECT * FROM kv@{NO_INDEX_JOIN}`.

This patch extends this mechanism with a new index hint: `NO_SCAN`
When specifies, this causes the query to fail with an error if the
query planner decides to issue a full table scan on that table.

For example:
```
root@:26257/t> SELECT * FROM t.kv@{NO_SCAN} where k%2 = 1;
pq: query plan requested full scan of table kv with NO_SCAN specified
```

This feature can be e.g. used to assert proper use of indexes, or
troubleshoot during manual query optimization, in a way slightly more
direct than inspecting the query plan with EXPLAIN every time.

Fixes #13969.
cc @scode

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13994)

### CODE REVIEW

**Comment by dt** [comment]
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/scan.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994#-KejH8KyMj2AcnEdSYRE:-KejH8KyMj2AcnEdSYRF:b-15gvj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d851cb7bb2a0f13fe00594c65301782a83bff08/pkg/sql/scan.go#L155)):*
> ```Go
> func (n *scanNode) initScan() error {
> 	if n.assertNoScan {
> 		if len(n.spans) == 0 || (len(n.spans) == 1 && sqlbase.PrettySpan(n.spans[0], 2) == "-") {
> ```

:dog: but is there a way to do this without comparing to a pretty-printed span?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994)*

**Comment by knz** [comment]
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/scan.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994#-KejH8KyMj2AcnEdSYRE:-KejHMfYMZ-tzsThrx1Q:b86b0p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d851cb7bb2a0f13fe00594c65301782a83bff08/pkg/sql/scan.go#L155)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

:dog: but is there a way to do this without comparing to a pretty-printed span?
</blockquote></details>

yes, just found it. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994)*

**Comment by dt** [comment]
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/scan.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994#-KejH8KyMj2AcnEdSYRE:-KejIJy2I-9a_yWlJeCU:b-ma32fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d851cb7bb2a0f13fe00594c65301782a83bff08/pkg/sql/scan.go#L155)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

yes, just found it. PTAL.
</blockquote></details>

still have no idea what i'm doing here, but looks better to me now :p I think IndexSpan has to alloc a []byte though, which is a little unfortunate just to do the comparison, but I have no idea how much that matters in practice.

---

*[pkg/sql/testdata/select_index_hints, line 228 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994#-KejJ8RSfddaIC2nrxPt:-KejJ8RSfddaIC2nrxPu:b7fkjbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/98fee898bc8bbca99e6c7cbf3a822e3223851a7a/pkg/sql/testdata/select_index_hints#L228)):*
> ```
> 
> query error query plan requested full scan of abcd@b with NO_SCAN specified
> SELECT * FROM abcd@{FORCE_INDEX=b,NO_SCAN} WHERE a >= 20 AND a <= 30
> ```

I'd include a positive test as well -- i.e. one that succeeds --  that includes a NO_SCAN.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994)*

**Comment by knz** [comment]
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/scan.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994#-KejH8KyMj2AcnEdSYRE:-KejWpiRgB54KWkaLQUp:bi91z7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d851cb7bb2a0f13fe00594c65301782a83bff08/pkg/sql/scan.go#L155)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

still have no idea what i'm doing here, but looks better to me now :p I think IndexSpan has to alloc a []byte though, which is a little unfortunate just to do the comparison, but I have no idea how much that matters in practice.
</blockquote></details>

Yeah we don't really have an easy way to mark the fact there was a full table scan.

---

*[pkg/sql/testdata/select_index_hints, line 228 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994#-KejJ8RSfddaIC2nrxPt:-KejWpJk360lr3ReQWjT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98fee898bc8bbca99e6c7cbf3a822e3223851a7a/pkg/sql/testdata/select_index_hints#L228)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'd include a positive test as well -- i.e. one that succeeds --  that includes a NO_SCAN.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13994)*

**Review by dt — APPROVED** [review]

**Comment by knz** [comment]
So I've thought about this some more and I have come independently to the same conclusion as @andreimatei : perhaps this feature as-is is a bit too half-baked.

The particular technical question that leads me to this conclusion is *"how to detect whether a query is performing a full table scan?"* 

As of today we do not have a straightforward, cheap way to answer this question. In many cases a full scan is indicated by a single span for the entire table, however in the presence of partitioned tables we can have multiple spans which amount to a "full scan" only in aggregate, not when looked at individually.

Also even with a single span we'd need to decode the span keys after they've been generated to determine whether they cover the entire table, and this decoding is non trivial either. The current code re-generates (i.e. re-allocates in memory)  a byte array every time this is tested, which is obviously poor design. 

What's needed is a boolean in the `Span` struct to annotate "full spans" in addition to the keys, but that would be a significant refactoring with considerations beyond the scope of what I can foresee myself.

In other words, "building this right" is not simple and I'm probably not doing it right in this PR.

**Comment by dt** [comment]
(configurable) logging and warning on full table scans or slow queries in general is another approach i've seen to this problem

**Comment by knz** [comment]
The approach was flawed, so closing this PR.
